### ISSUE TITLE
Eval is not respecting `nthread` parameter

### ISSUE THREAD

**Issue by s-banach** [dmlc/xgboost#8802]
My code currently looks like this:
```python
xgb.train(
    params=dict(
        objective="binary:logistic",
        eval_metric="auc",
        tree_method="exact",
        learning_rate=0.02,
        nthread=2,
    ),
    dtrain=d_train,
    evals=[(d_test, "Test")],
    num_boost_round=1000,
    callbacks=[callback.EarlyStopping(rounds=early_stopping_rounds, maximize=True)],
    verbose_eval=False,
)
```

If I remove `evals` and `callbacks`, I get 200% CPU usage as desired.  
With them present, I'm seeing essentially uncapped CPU usage.

My guess? Perhaps the `booster.predict()` method does not respect the `nthread` parameter.  
Is there a recommended workaround? Maybe I have to defined a custom eval function that sets the `nthread` parameter on `booster.predict`.

**Comment by s-banach** [comment]
Maybe the `auc` metric is calling some BLAS-type library that uses unlimited cores.  
I can wrap `xgb.train()` in a thread pool controller from `threadpoolctl` and it seems to work.

**Comment by trivialfis** [comment]
Thank you for raising the issue, it's actually caused by calling the parallel sorting algorithms from gcc or msvc. It's fixed by this PR: https://github.com/dmlc/xgboost/pull/8735 .
